src/theory/tools/wp/bir_htScript.sml

```diff
@@ -0,0 +1,614 @@
+]
+);
+
+(* TODO: Move this elsewhere... *)
+val bir_exec_stmtsB_exists =
+  store_thm("bir_exec_stmtsB_exists",
+  ``!stmtsB l c st.
+      ?l' c' st'.
+        bir_exec_stmtsB stmtsB (l,c,st) = (l', c', st')``,
+
+Induct_on `stmtsB` >- (
+  FULL_SIMP_TAC std_ss [bir_exec_stmtsB_def]
+) >>
+REPEAT STRIP_TAC >>
+FULL_SIMP_TAC std_ss [bir_exec_stmtsB_def] >>
+Cases_on `bir_state_is_terminated st` >- (
+  FULL_SIMP_TAC std_ss []
+) >>
+ASSUME_TAC (SPECL [``h:'a bir_stmt_basic_t``, ``st:bir_state_t``]
+                  bir_exec_stmtB_exists) >>
+FULL_SIMP_TAC std_ss [LET_DEF]
+);
+
+val bir_block_not_assume_never_assviol =
+  store_thm("bir_block_not_assume_never_assviol",
+  ``!prog bl st l' c' st'.
+    bir_block_has_no_assumes bl ==>
+    (st.bst_status <> BST_AssumptionViolated) ==>
+    (bir_exec_block prog bl st = (l', c', st')) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+FULL_SIMP_TAC std_ss [bir_block_has_no_assumes_def,
+                      bir_exec_block_def] >>
+REPEAT STRIP_TAC >>
+IMP_RES_TAC bir_stmtsB_not_assume_never_assviol >>
+subgoal `?l_test c_test st'_test.
+	   bir_exec_stmtsB bl.bb_statements ([],0,st) =
+             (l_test, c_test, st'_test)` >- (
+  FULL_SIMP_TAC std_ss [bir_exec_stmtsB_exists]
+) >>
+FULL_SIMP_TAC std_ss [LET_DEF] >>
+PAT_X_ASSUM ``!st' l' l c' c. blah``
+	    (fn thm => ASSUME_TAC
+	      (SPECL [``st'_test:bir_state_t``,
+		      ``l_test: 'a list``,
+                      ``[]:'a list``,
+		      ``c_test:num``,
+		      ``0:num``] thm
+	      )
+	    ) >>
+REV_FULL_SIMP_TAC std_ss [] >>
+Cases_on `bir_state_is_terminated st'_test` >| [
+  Cases_on `st'_test` >>
+  Cases_on `st` >>
+  Cases_on `st'` >>
+  FULL_SIMP_TAC (std_ss++holBACore_ss)
+                [bir_state_t_fn_updates, bir_state_t_bst_status,
+                 bir_state_t_bst_pc],
+
+  FULL_SIMP_TAC (std_ss++holBACore_ss)
+                [bir_state_t_fn_updates, bir_state_t_bst_status,
+                 bir_state_t_bst_pc] >>
+  Cases_on `bir_state_is_terminated
+              (bir_exec_stmtE prog bl.bb_last_statement
+                 st'_test)` >> (
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [] >>
+    Cases_on `st'` >>
+    ASSUME_TAC (SPECL [``prog:'a bir_program_t``,
+                       ``bl.bb_last_statement:bir_stmt_end_t``,
+                       ``st'_test:bir_state_t``]
+                      bir_exec_stmtE_not_assviol
+               ) >>
+    Q.ABBREV_TAC `st' = bir_exec_stmtE prog bl.bb_last_statement
+                                       st'_test` >>
+    Cases_on `st'` >>
+    REV_FULL_SIMP_TAC (std_ss++holBACore_ss)
+                      [bir_state_t_fn_updates]
+  )
+]
+);
+
+val bir_prog_has_no_assumes_def = Define `
+  (bir_prog_has_no_assumes (BirProgram []) = T) /\
+  (bir_prog_has_no_assumes (BirProgram (h::t)) =
+    (bir_block_has_no_assumes h) /\
+    (bir_prog_has_no_assumes (BirProgram t))
+  )
+`;
+
+val test_lemma_3 = store_thm("test_lemma_3",
```

Comment:
Should this be stored in the theory? (`store_thm` => `prove` ?)

Reply:
I have now renamed it to fit the terminology of `listTheory` where, IMO, this lemma should ideally be found. Yes, I am quite sure this should be stated with `store_thm`, but not in this theory.

@andreaslindner, what does the textbook say when we prove stuff that should really go into HOL4 mainline theories? As for this particular one, I suggest I drop it in `bir_auxiliaryScript.sml` (IMO this also applies to `FEVERY_FEVERY_DRESTRICT_thm` in `bir_wpTheory`, which I didn't author).

Reply:
I was just saying this because of the name of the theorem.

Reply:
yes, maybe we should collect them in auxiliary and if you feel like it you mainline it and we could potentially remove it from auxiliary later at some point. I don't think we should develop more workflow for this than that :)